Do not allow splits within rows.

petermattis opened the pull request:
Append the column ID length to SQL generated keys. Added
keys.MakeSplitKey() which strips off the column ID from SQL table
keys. Utilize from within AdminSplit to prevent splitting within an SQL
row.

Changed the type of client.KeyValue.Key to roachpb.Key so that the Key
field pretty-prints without extra effort. Similarily, changed the type
of sql.indexEntry.key to roachpb.Key.

Fixes #3461.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3513)

<!-- Reviewable:end -->

petermattis committed:
Do not allow splits within rows.

Append the column ID length to SQL generated keys. Added
keys.MakeSplitKey() which strips off the column ID from SQL table
keys. Utilize from within AdminSplit to prevent splitting within an SQL
row.

Changed the type of client.KeyValue.Key to roachpb.Key so that the Key
field pretty-prints without extra effort. Similarily, changed the type
of sql.indexEntry.key to roachpb.Key.

petermattis commented:
I plan to add some additional tests of `AdminSplit` to `storage/client_split_test.go` before merging.

petermattis commented on storage/client_split_test.go:
```diff
@@ -85,13 +85,13 @@ func TestStoreRangeSplitAtIllegalKeys(t *testing.T) {
 	defer stopper.Stop()
 
-	key := keys.TableDataMin
+	key := keys.UserTableDataMin
```
I'm not quite sure why this was working before. `keys.TableDataMin` lies in the system DB span and is not a valid split key.

tschottdorf commented on roachpb/data.go:
```diff
@@ -784,6 +784,11 @@ func (s Span) Equal(o Span) bool {
 }
 
+// ContainsKey returns whether this span contains the specified key.
+func (s Span) ContainsKey(key Key) bool {
```
I would advise against having comparison functions on `Key` types because of local addressing. If you're trying to use `ContainsKey`, chances are that you want to be looking at an `RSpan`.

petermattis replied:
This is used from `engine.IsValidSplitKey` which was previously comparing on keys directly. Happy to move the comparisons back to that routine if it is too dangerous to leave here. But I'm not clear on how I can operate on `RKeys` at that level.

tamird commented:
Reviewed 14 of 16 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[keys/keys.go, line 398 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6UNK691-gLembPbmoJ:-K6UNK691-gLembPbmoK:426919849) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad0f7ec0091acd3092d77b430e7640ea5aeed7e9/keys/keys.go#L398)):*
this is a bit non-obvious; can you use a local for `buf[n-1:]` and then `if int(colIDLen) > len(buf[n-1:])` here? slightly simplifies the final slice below, too.

---

*[keys/keys.go, line 399 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6UM_3K7_gd9hzHkEYo:-K6UM_3K7_gd9hzHkEYp:-1907503976) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad0f7ec0091acd3092d77b430e7640ea5aeed7e9/keys/keys.go#L399)):*
not an error?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[keys/keys.go, line 398 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6UNK691-gLembPbmoJ:-K6UnhTXgs4SX3D_mVug:-1574613657) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad0f7ec0091acd3092d77b430e7640ea5aeed7e9/keys/keys.go#L398)):*
Not sure how that would help the final slice below as we're slicing into `key`, not `buf`. I've expanded the comments here. 

---

*[keys/keys.go, line 399 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6UM_3K7_gd9hzHkEYo:-K6UnppsL43X4By_MeqP:1078909596) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad0f7ec0091acd3092d77b430e7640ea5aeed7e9/keys/keys.go#L399)):*
There are table keys passed to this routine which do not have a column ID length suffix. In particular, keys generated by `SystemConfig.ComputeSplitKeys`. I suppose I could fix the keys generated there, but it seems easier to just be permissive of the keys passed in to this routine.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[keys/keys.go, line 361 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6_7BlzRzACXmFO8R-o:-K6_7BlzRzACXmFO8R-p:-10770704) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e3156e7e6838b22997dbc34fc634dc3c11eade/keys/keys.go#L361)):*
Can we just append an extra zero byte to the keys constructed in `ComputeSplitKeys` to avoid a special case here? I don't think we care about user-constructed keys at this level; a future SQL-aware admin split tool could construct well-formed keys. 

---

*[sql/keys.go, line 105 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6_9FDeinyt2Um2LaAk:-K6_9FDeinyt2Um2LaAl:-503554344) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e3156e7e6838b22997dbc34fc634dc3c11eade/sql/keys.go#L105)):*
Isn't this last field supposed to be the size of the column ID portion, not the base/PK portion?

It's a little misleading to use `EncodeUvarint` here since `stripColumnIDLength` will fail if it is more than one byte. That can't happen since the length of a varint can always fit in a single byte but it would be good to add a comment here to that effect (or bypass the varint encoding).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[keys/keys.go, line 361 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6_7BlzRzACXmFO8R-o:-K6_TpNJFJr0gTeR83ff:1285650113) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e3156e7e6838b22997dbc34fc634dc3c11eade/keys/keys.go#L361)):*
Let me investigate. The current `cockroach split` command will split ranges in tables if given correctly encoded keys. I suppose we should make some (all?) of the cli commands more SQL aware.

---

*[sql/keys.go, line 105 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6_9FDeinyt2Um2LaAk:-K6_TN3maCC0QFB_nO-b:-1193286887) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e3156e7e6838b22997dbc34fc634dc3c11eade/sql/keys.go#L105)):*
Yeah, that is supposed to be `len(key) - size`. Clearly the test coverage here is weak. I'll add a comment about the length portion being guaranteed to fit in a single byte when varint encoded. I originally started with just storing the length directly without varint encoding, but that caused pretty-printing to get upset and trying to fix that seemed more troublesome than just using a varint encoding here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 8 of 17 files reviewed at latest revision, 3 unresolved discussions.

---

*[keys/keys.go, line 361 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6_7BlzRzACXmFO8R-o:-K6abIAyooVbOpNC2RUc:-2078499010) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e3156e7e6838b22997dbc34fc634dc3c11eade/keys/keys.go#L361)):*
After changing `ComputeSplitKeys` to encode the split keys with an encoded column ID suffix size there are some weird test failures in the `storage` split tests which I've been unable to track down. You can see these by changing the code at `config.go:315-316`. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 8 of 18 files reviewed at latest revision, 3 unresolved discussions.

---

*[keys/keys.go, line 361 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6_7BlzRzACXmFO8R-o:-K6dMTQUD2Y2xZeGbqlI:758073733) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e3156e7e6838b22997dbc34fc634dc3c11eade/keys/keys.go#L361)):*
Just needed fresh eyes. Found the problem in a matter of minutes this morning (had forgotten to change `MakeSplitKey` to be more permissive). Tests are all passing now, though I still want to add one that tests `AdminSplit` trying to split within a row. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 8 of 18 files reviewed at latest revision, 6 unresolved discussions.

---

*[keys/keys.go, line 353 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6cuT2Jjk4uasYSpYlA:-K6cuT2Jjk4uasYSpYlB:598526088) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fff06328022f2bbaa2bbf3c1fb2eeccfb51c7c/keys/keys.go#L353)):*
the comment suggests that no copy of `rowKey` is made.

---

*[keys/keys.go, line 368 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6cuqJaPP9dtge5e353:-K6cuqJaPP9dtge5e354:-313461301) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fff06328022f2bbaa2bbf3c1fb2eeccfb51c7c/keys/keys.go#L368)):*
so it skips the column id? Comparing to `EncodeColumnKey`, this looks more like a column key with 0 ID and missing size suffix.

---

*[keys/keys.go, line 373 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6cveaFndnFt_0TrGnr:-K6cveaFndnFt_0TrGns:-1011354851) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fff06328022f2bbaa2bbf3c1fb2eeccfb51c7c/keys/keys.go#L373)):*
If this code is to stay (looking at the discussion below) some schematic examples would be helpful. The code below is nontrivial 

---

*[roachpb/data.go, line 788 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6HL7ThOQQsuLty2pp6-r1-788:-K6csBylTAEM-1e9O4I-:133296533) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c7ac61647ceaec72fa3159487a832bcf6847c7/roachpb/data.go#L788)):*
yeah, doing it ad-hoc is probably the better option here then. The thing we want to avoid is give the impression that `roachpb.Key` have an ordering. We made sure that wasn't going to happen when we introduced range-local addressing. When you're looking for a split key, you should always make sure that the key isn't a range-local key anyways, so you could take `resolvedKey := keys.Addr(splitKey)` and if `!bytes.Equal(resolvedKey, splitKey)` then you want to declare the split invalid anyway. This is probably covered by the illegal split intervals implicitly, though.

When I made the change which introduced `RKey`, I really should've introduced a new key type for **unresolved** keys (not for resolved keys, `RKey`). Everything that happens at the MVCC layer should in fact be an `RKey`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 7 of 18 files reviewed at latest revision, 6 unresolved discussions.

---

*[keys/keys.go, line 353 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6cuT2Jjk4uasYSpYlA:-K6dXoPaZoc6AUHN8xqo:-1294528301) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fff06328022f2bbaa2bbf3c1fb2eeccfb51c7c/keys/keys.go#L353)):*
Yeah, I tried not making a copy, but it turns out we need to here. I've renamed this method back to `MakeColumnKey` and clarified the comment.

---

*[keys/keys.go, line 368 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6cuqJaPP9dtge5e353:-K6dYfuCeozgfxnmAQ5R:776175734) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fff06328022f2bbaa2bbf3c1fb2eeccfb51c7c/keys/keys.go#L368)):*
Yes, it skips encoding the column ID and only encodes the number of bytes of encoded column ID (0 bytes).

---

*[keys/keys.go, line 373 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6cveaFndnFt_0TrGnr:-K6dYn2byXJtKm63LxvX:1135947138) ([raw file](https://github.com/cockroachdb/cockroach/blob/60fff06328022f2bbaa2bbf3c1fb2eeccfb51c7c/keys/keys.go#L373)):*
This code has been simplified. 

---

*[roachpb/data.go, line 788 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6HL7ThOQQsuLty2pp6-r1-788:-K6dZN9-egwj3dITyea8:-1525454147) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c7ac61647ceaec72fa3159487a832bcf6847c7/roachpb/data.go#L788)):*
The `Key` vs `RKey` stuff was confusing as I was making this change, perhaps because `MVCC` deals in `roachpb.Key` while most stuff at the in `storage` deals with `roachpb.RKey`. Outside of the scope of this PR to fix that. Care to file a bug?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 4 files at r5, 1 of 7 files at r6.
Review status: 10 of 18 files reviewed at latest revision, 8 unresolved discussions.

---

*[config/config.go, line 313 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6e9EesOsImqEBDHf0R:-K6e9EesOsImqEBDHf0S:-734662450) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/config/config.go#L313)):*
TODO should go.

---

*[keys/keys.go, line 384 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eBBvyxfw6g0hgmrgt:-K6eBBvyxfw6g0hgmrgu:-904907157) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/keys/keys.go#L384)):*
... which means it's again not a real table key?

---

*[storage/client_split_test.go, line 148 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eEsJwJo5ZOo_ZyCw-:-K6eEsJwJo5ZOo_ZyCw0:-1228438948) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/storage/client_split_test.go#L148)):*
does this test correspond to a table like
`CREATE TABLE x (id string, col1 something, col2 something)`? Worth a small comment.

---

*[storage/client_split_test.go, line 159 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eDlyI6DWy7rCKkyeT:-K6eDlyJDRarpjU6T2K-:-1606993829) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/storage/client_split_test.go#L159)):*
wouldn't splitting at `col1Key` put separate into `[something, col1Key)` and `[col1Key, rest)` anyway, so there's no danger here of separating the two? Looks to me that it ought to be`col2Key`. And do we still care about user keys? `MakeSplitKey` looks like it might make a mess of things if you start with `append(col1Key, '\x00\xff')` and just split right there (?).

---

*[storage/client_split_test.go, line 179 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eEYi6VpPOPRVmuSD0:-K6eEYi6VpPOPRVmuSD1:-454108749) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/storage/client_split_test.go#L179)):*
that just verifies that `(left range).EndKey = (right range).StartKey` after splitting. Doesn't seem to be adding anything to the test?

---

*[roachpb/data.go, line 788 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6HL7ThOQQsuLty2pp6-r1-788:-K6eAq9g3rcDUMVmo5le:-2012836951) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c7ac61647ceaec72fa3159487a832bcf6847c7/roachpb/data.go#L788)):*
filed #3541.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 9 files at r4, 2 of 4 files at r5, 7 of 7 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[config/config_test.go, line 274 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eGR1OdbagVPzv5_1W:-K6eGR1OdbagVPzv5_1X:114711086) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/config/config_test.go#L274)):*
what are these changes about, and where do these magic numbers come from?

---

*[keys/keys.go, line 355 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eHHV_QcRYoYVXnuMN:-K6eHHVaDTVt0C5_uVHd:1252818242) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/keys/keys.go#L355)):*
nit: isn't the style usually used `key := append([]byte(nil), rowKey...)`?

---

*[keys/keys_test.go, line 360 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eGh5OY4dKObL457Pl:-K6eGh5OY4dKObL457Pm:-475482483) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/keys/keys_test.go#L360)):*
`%x`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 15 of 18 files reviewed at latest revision, 10 unresolved discussions.

---

*[config/config.go, line 313 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6e9EesOsImqEBDHf0R:-K6eNSOQY3Ug_JS1E8kC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/config/config.go#L313)):*
Done.

---

*[config/config_test.go, line 274 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eGR1OdbagVPzv5_1W:-K6eNUiRlNUGyuuhPwq8:-2124574523) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/config/config_test.go#L274)):*
The test is a bit opaque (I'm dealing with what was already here). `allUserSplits` is initialized above and what I changed was the expected splits. The expected splits changed because the split key (`/Table/1005/0) is before the end key (`/Table/1005/"bar"`). Prior to this change, the split key was `/Table/1005` which would be before the end key. 

---

*[keys/keys.go, line 355 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eHHV_QcRYoYVXnuMN:-K6eOAdFeQ_74g5kflwP:-1488412297) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/keys/keys.go#L355)):*
Yeah, this was copy&paste from elsewhere. I'll change.

---

*[keys/keys.go, line 384 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eBBvyxfw6g0hgmrgt:-K6eOQtHZk6LFiHgYSwN:-204418411) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/keys/keys.go#L384)):*
Yeah. I suppose I could return an error now. This shouldn't happen. Let me see if that breaks anything. Nope. Changed.

---

*[keys/keys_test.go, line 360 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eGh5OY4dKObL457Pl:-K6eN7lRErJsjllDJJD6:-70359245) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/keys/keys_test.go#L360)):*
Print as hex with spaces between each byte.

---

*[storage/client_split_test.go, line 148 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eEsJwJo5ZOo_ZyCw-:-K6eOgdSH_X59X3ukUK0:486066136) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/storage/client_split_test.go#L148)):*
Something like that. I'll add a comment.

---

*[storage/client_split_test.go, line 159 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eDlyI6DWy7rCKkyeT:-K6eOpmhOOyjWGJc3LEf:1500307347) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/storage/client_split_test.go#L159)):*
Prior to this change, splitting at `col1Key` would create `[something, col1Key]` and `[col1Key, rest]`. Not it creates `[something, rowKey]` and `[rowKey, rest]`. Splitting on `col2Key` should have the same effect (i.e. we split at the row key, not inside the row). 

What user keys are you thinking of? All user data is in table keys.

---

*[storage/client_split_test.go, line 179 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eEYi6VpPOPRVmuSD0:-K6eNIoVlf-EjN_gVaQJ:1070502055) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/storage/client_split_test.go#L179)):*
Yeah, it adds little, though it does verify that `tableKey` is in a different range than `rowKey` after the split, so I'd like to leave it in.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[config/config_test.go, line 274 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eGR1OdbagVPzv5_1W:-K6eQ7J5rSPokqLwKXFL:1776681207) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/config/config_test.go#L274)):*
I see.

---

*[keys/keys_test.go, line 360 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eGh5OY4dKObL457Pl:-K6eQGXfd8dKQLwzGPMW:1436666210) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/keys/keys_test.go#L360)):*
right, current code has a space between `%` and `x`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[keys/keys_test.go, line 360 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eGh5OY4dKObL457Pl:-K6eQmnPy3KvhZaTKGhV:-1433330946) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/keys/keys_test.go#L360)):*
errr, disregard. TIL

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[config/config_test.go, line 274 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eGR1OdbagVPzv5_1W:-K6eR-iorFXQrBIK4FZv:1325043038) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/config/config_test.go#L274)):*
Yeah, the test isn't exactly pretty. 

---

*[keys/keys_test.go, line 360 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eGh5OY4dKObL457Pl:-K6eQkL3frvVboabISTp:-1205173953) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/keys/keys_test.go#L360)):*
Yep, the space is intentional: http://play.golang.org/p/uhQZMMvyh1

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 17 of 18 files reviewed at latest revision, 7 unresolved discussions.

---

*[keys/keys.go, line 363 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eXmsYse-DNj3FLAVU:-K6eXmsZkG3FReeL_sKU:1075199120) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653be72b16ea8dba8735fb3dd3ff987c40f1fa9/keys/keys.go#L363)):*
I find the name "sentinel key" pretty vague, and the way this function is used in `ComputeSplitKeys` isn't really a sentinel anyway (at least not the way row sentinel keys are). How about `MakeNonColumnKey`?

---

*[keys/keys.go, line 401 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eWLZ1ZO2B4I-kkRmz:-K6eWLZ1ZO2B4I-kkRn-:659458317) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653be72b16ea8dba8735fb3dd3ff987c40f1fa9/keys/keys.go#L401)):*
Is this still true? I'd like for this to be an error.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-:-K6eZLsWM_EdY8SjAwT9:676834233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 15 of 18 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/client_split_test.go, line 159 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eDlyI6DWy7rCKkyeT:-K6eZcizbt-cs6_mxN8u:-1188599541) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/storage/client_split_test.go#L159)):*
I'm thinking `./cockroach range split <the key i cooked up above>`. Nothing that would be picked automatically (i.e. not an existing key). Basically asking whether we plan to allow manual override of this logic. If not, we should error accordingly. I think it connects with Ben's comment about erroring out in more cases.

Would you mind splitting on `col2Key` instead? The comment  on 171 is confusing as is. Or take out 171-174 - looks like 175ff already establishes something stronger.

---

*[storage/client_split_test.go, line 155 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6ed2gKQ1ZqEqfISEVY:-K6ed2gKQ1ZqEqfISEVZ:175169565) ([raw file](https://github.com/cockroachdb/cockroach/blob/5872a37fe841b904ea6907e43636f8a777de2854/storage/client_split_test.go#L155)):*
probably will never matter, but we've already had to deal with cruft from putting "any old thing" in various places - if there's an easy way to write valid data, do that please (if a sql-sensitive queue comes along and tries to decode, it'll barf).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 3 files at r7.
Review status: 15 of 18 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[keys/keys.go, line 363 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eXmsYse-DNj3FLAVU:-K6e_hLqmbTSM2bE8T5O:1314565758) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653be72b16ea8dba8735fb3dd3ff987c40f1fa9/keys/keys.go#L363)):*
Yeah, `Sentinel` isn't a great term here. But `NonColumn` is unwieldy. Let me think about this over dinner.

---

*[keys/keys.go, line 401 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eWLZ1ZO2B4I-kkRmz:-K6eZm_xvQ2IATsFKNEE:-778169487) ([raw file](https://github.com/cockroachdb/cockroach/blob/a653be72b16ea8dba8735fb3dd3ff987c40f1fa9/keys/keys.go#L401)):*
Not necessary (after a few test fixes). Error returned.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/client_split_test.go, line 159 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6eDlyI6DWy7rCKkyeT:-K6jV_ITGvTdKdSFSJ7v:-1230589809) ([raw file](https://github.com/cockroachdb/cockroach/blob/363d626c505696a2ac2b630ccd836cc82982eb2f/storage/client_split_test.go#L159)):*
Sure, changed to splitting on `col2Key`. I get confused that splitting on a key splits before the key, not after the key.

`./cockroach range split` needs rethinking per #3529. Currently, we'll error out if the user supplies on an invalid table key that they just cooked up. That seems reasonable for now, but longer term I don't think we want to require the user generate the full KV key but instead specify a table name and primary key values for that table to split on.

---

*[storage/client_split_test.go, line 155 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513#-K6ed2gKQ1ZqEqfISEVY:-K6jV1F-gAdkTVAMlOSJ:1669712401) ([raw file](https://github.com/cockroachdb/cockroach/blob/5872a37fe841b904ea6907e43636f8a777de2854/storage/client_split_test.go#L155)):*
This is actually valid SQL data for a `STRING` column.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3513)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

tschottdorf commented:
Haven't had a chance to look over the latest changes, but don't let me hold
you up.

On Wed, Dec 30, 2015, 00:43 Ben Darnell <notifications@github.com> wrote:

> LGTM
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/3513#issuecomment-167900724>
> .
>
